pkg/apis/tenancy/v1alpha1/types_clusterworkspace.go

```diff
@@ -0,0 +1,260 @@
+// +kubebuilder:printcolumn:name="Type",type=string,JSONPath=`.spec.type.name`,description="Type of the workspace"
+// +kubebuilder:printcolumn:name="URL",type=string,JSONPath=`.status.baseURL`,description="URL to access the workspace"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+type ClusterWorkspace struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec ClusterWorkspaceSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status ClusterWorkspaceStatus `json:"status,omitempty"`
+}
+
+// ClusterWorkspaceSpec holds the desired state of the ClusterWorkspace.
+type ClusterWorkspaceSpec struct {
+	// +optional
+	ReadOnly bool `json:"readOnly,omitempty"`
+
+	// type defines properties of the workspace both on creation (e.g. initial
+	// resources and initially installed APIs) and during runtime (e.g. permissions).
+	// If no type is provided, the default type for the workspace in which this workspace
+	// is nesting will be used.
+	//
+	// The type is a reference to a WorkspaceType in the listed workspace, but
+	// lower-cased. The WorkspaceType existence is validated at admission during
+	// creation. The type is immutable after creation. The use of a type is gated via
+	// the RBAC workspacetypes/use resource permission.
+	//
+	// +optional
+	Type WorkspaceTypesReference `json:"type,omitempty"`
+
+	// shard constraints onto which shards this cluster workspace can be scheduled to.
+	// if the constraint is not fulfilled by the current location stored in the status,
+	// movement will be attempted.
+	//
+	// Either shard name or shard selector must be specified.
+	//
+	// If the no shard constraints are specified, an arbitrary shard is chosen.
+	//
+	// +optional
+	Shard *ShardConstraints `json:"shard,omitempty"`
+}
+
+type ShardConstraints struct {
+	// name is the name of Shard.
+	//
+	// +optional
+	// +kubebuilder:validation:Pattern=`^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$`
+	Name string `json:"name,omitempty"`
+
+	// selector is a label selector that filters shard scheduling targets.
+	//
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+}
+
+// WorkspaceTypesReference is a globally unique, fully qualified reference to a
+// cluster workspace type.
+type WorkspaceTypesReference struct {
+	// name is the name of the WorkspaceType
+	//
+	// +required
+	// +kubebuilder:validation:Required
+	Name WorkspaceTypesName `json:"name"`
+
+	// path is an absolute reference to the workspace that owns this type, e.g. root:org:ws.
+	//
+	// +optional
+	// +kubebuilder:validation:Pattern:="^[a-z0-9]([-a-z0-9]*[a-z0-9])?(:[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
+	Path string `json:"path,omitempty"`
+}
+
+// WorkspaceTypesName is a name of a WorkspaceType
+//
+// +kubebuilder:validation:Pattern=`^[a-z]([a-z0-9-]{0,61}[a-z0-9])?`
+type WorkspaceTypesName string
```

Comment:
Should this be singular - WorkspaceTypeName?

Reply:
done.

Had had a little rename accident. Hope have now fixed all plural occurances.